### ISSUE TITLE
Primary conditions take precedence over derived ones

### ISSUE THREAD

**Issue by leklachu** [endless-sky/endless-sky#7217]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Playing through FW campaign, chose Reconciliation, went to rescue Sawyer... and lo and behold, I'm friends with all the Syndicate ships! Actually, my reputation is 481.21, and unaffected by the mission action `"reputation: Syndicate" = -1000`

### Steps to Reproduce

1. Play through the FW campaign
2. Choose to rescue the defector
3. Go to Alioth and watch while your FW escorts fight the Syndicate, but you look on and wonder.

Or: Load save previous-2 (from below), continue to Tarazed, and find the same.

### Expected Behavior

3. Go to Alioth and be hostile to the Syndicate.

### Link to save file

[Diana Thudey.txt](https://github.com/endless-sky/endless-sky/files/9489835/Diana.Thudey.txt)
[Diana Thudey~~previous-1.txt](https://github.com/endless-sky/endless-sky/files/9489837/Diana.Thudey.previous-1.txt)
[Diana Thudey~~previous-2.txt](https://github.com/endless-sky/endless-sky/files/9489838/Diana.Thudey.previous-2.txt)

Three saves: before/on/after Tarazed

### Operating System

OpenSUSE Tumbleweed

### Game Source

Built from source

### Game Version

commit SHA ac6e0ac4f9346a59108e8865debf0cd687a0a313

### Additional Information

This has been fine in fairly-recent playthroughs (few months ago?). Tried disabling all plugins - still the same.

Everything seems to be right in the files.
- Even the reputation-change action is there in the mission as recorded briefly in the savefile. (but the reputation remains constant)
- The other actions in that block: `event: fw at war wtith Syndicate` and `event: battle of alioth` both trigger

I haven't tried replicating from start. If no one figures out what's up, I'll try and find a moment to do a speedrun from scratch to replicate.
- I have tried several times reloading from just before Tarazed (FW Rand 1B: active), and the same happens.

**Comment by Amazinite** [comment]
The reason that SetReputationConditions and CheckReputationConditions existed is because in order for a ConditionSet to access information, that information had to be in the form of a Condition. Those two function had the responsibility of duplicating changes between the actual reputation values and the condition reputation values. What #5244 did was remove the need to duplicate information into the conditions by allowing information to be requested on demand as a condition, hence removing the need for those two functions to duplicate the information.

So it's not that those functions are gone that's causing this, it's that there's something wrong with the new set function for reputation.

https://github.com/endless-sky/endless-sky/blob/12f70efe38c3ac6e4127e8b586e6414b46ab8031/source/PlayerInfo.cpp#L2682-L2690

**Comment by Amazinite** [comment]
Should we be swapping the prefix match with the exact name match?

https://github.com/endless-sky/endless-sky/blob/12f70efe38c3ac6e4127e8b586e6414b46ab8031/source/ConditionsStore.cpp#L450-L472

**Comment by leklachu** [comment]
If I scrub just the `reputation: Syndicate` conditioni from my savefile, the mission just re-adds that. If I scrub all `reputation: _` from my file, then the `"reputation with"` section updated for the Syndicate properly.

**Comment by leklachu** [comment]
I think I've found it.

So, a GetEntry (ConditionsStore.cpp L450) will give a pointer either to the exact name we're looking for, or to a block of provider-provided entries (e.g. "reputation: __"s ). If there's even one "reputation: __" old-style condition, then that block is returned, and the next logic deals with that.

GetEntry is called by ConditionsStore::Set (L329). That asks GetEntry for the entry, but if that returns nothing, i.e. no "rep: __"s at all, then it fires off `(storage[name]).value = value;`, which I presume is setting the value correctly.
- otherwise it asks is the entry not a provider, then sets value; or-then finally sets via the provider

In which case, we need a check that a value *is* special rather than *isn't* a normal condition. Perhaps that's what you meant by
> Or should we scrub the primary conditions of anything that matches a derived one, prefixed included?

\- I haven't fully got my head round what primary and derived really mean here yet.

I guess that makes sense with your WorldForge tests too. ConditionsStore::Has uses the same logic, so in a mission check, that's being called and finds the "rep: __" entry; but the logic for ships attacking in game must come from elsewhere.

**Comment by petervdmeer** [comment]
Thanks for filing the bug-report. If we have a "prefixed derived condition provider" for `reputation: `, then we should indeed remove (scrub) all `reputation: Syndicate`, `reputation: Republic` direct conditions (and maybe apply the values from those on the prefixed provider before removing them).

Looks like I forgot to add this functionality in the original PR.

I don't have a lot of time in the coming weeks (I just started with a new job), but I agree that this needs to be fixed for the 0.9.15 milestone.

**Comment by leklachu** [comment]
So, do you want the program to,

1. check the savefile for `reputation: ` and delete them every time
2. check once, clear `reputation: ` from the save, then write a `version` into the savefile - could be good for bug reports, and for any future changes that affect how the savefile works
3. have a special clause to ignore `reputation: ` conditions when reading/writing reputations, thereby using the `reputation with` savefile block
4. publish, either as an external program, or a button within the game, a "fix savefile" script - again, might be handy for certain bugs, including compatibility issues from content expansion, that typically get fixed by having an invisible in-game mission

I'll look more at the code whenever I have time, and maybe I can be some help.

**Comment by petervdmeer** [comment]
No, this should not be solved just for reputations, but is a more generic piece of functionality and should be solved for any condition that exists in the savegame but moved to derived conditions.

I feel we should start with a warning in `ConditionsStore::GetProviderPrefixed()` [around line 420](https://github.com/endless-sky/endless-sky/blob/68eab0a290499704c47491092404fa7dd6205b49/source/ConditionsStore.cpp#L420) when a prefixed-provider gets added while other conditions within the prefixed range already exist.

My assumption when I wrote the code was that the prefixed-providers are always setup before the actual loading of the conditions from the savegame (but this assumption is probably wrong, and I should have checked the assumption).


**For later, once the warning is in place:**
If the warning shows that issues exist, then we can also come up with a good way to solve the issues. Just erasing (scrubbing) is the simplest, but I expect that we can also perform a more smart recovery by applying the conditions on the prefixed-provider before erasing(scrubbing). (Such a more smart recovery helps to convert "old savegames" to "new savegames" when/if we move conditions from primary to derived.)

Note: If this "for later" text doesn't make sense, then please only add the warning in a PR. The warning will really help to make visible if a problem exists at the location where I expect it to be.

**Comment by leklachu** [comment]
> My assumption when I wrote the code was that the prefixed-providers are always setup before the actual loading of the conditions from the savegame (but this assumption is probably wrong, and I should have checked the assumption).

I think: `PlayerInfo.cpp`, `PlayerInfo::Load` L122, reads the parts of the file
- L173, when it finds `Token(0) == "reputation with"` loads the reputations into `GameData`
- L234, when it finds  `"conditions"`, loads the conditions directly into the ConditionsStore
- then after reading the file, L301 `RegisterDerivedConditions()` makes the GameData gvmt reputations available as derived conditions

---

So if I understand right: Both are loaded in, with the savefile conditions first.
- If they loaded after the derived conditions, each call to `Set` would find the derived prefix and act how we want (see more below)

What happens when the game tries to read/write `reputation: Syndicate` (for example),
- find the first entry alphabetically after `reputation: Syndicate`
  - this includes `reputation: __` entries and the derived `reputation: ` prefix
- steps back one.
  - if `reputation: Syndicate` exists, it'll match that as the condition to read/write
  - if any `reputation: __` alphabetically before Syndicate exist, it'll see the last one of those, see it doesn't match, and conclude there's no entry so make a new one
  - if there is no alphabetically-prior entry, the step-back will land on the `reputation: ` prefix (as the `name` in `storage`), and will match that, then use the set/get functions contained inside

I have tested this by deleting specifically all `reputation: __` conditions of Syndicate and before, but not the ones after. In that case, it works properly. If I just delete `reputation: Syndicate` it does the wrong behaviour.

---

In which case your warning + solution sounds like it would work

If it were possible to put `RegisterDerivedConditions()` above read-file... I haven't checked saving yet (and need to stop for now)
- loading would update the derived conditions from the savefile primary conditions
  - for values like reputation, would that lose precision? that the primary conditions are `int`s and `"reputation with"` holds fractions that I guess are floats
- if saving freshly writes down everything it knows, then the now-defunct primary conditions would just be forgotten. All's well
- if not, and they remain, then every new load would overwrite the proper list from the no-longer-updated old conditions

**Comment by leklachu** [comment]
I took a look through the logic to register derived conditions. Can't say I'm certain, but pretty confident now:

- The game datastructures to store reputations etc are created before load-ing or new-ing a player
- loading the savefile populates those structures with the actual data
- `RegisterDerivedConditions()` sets up functions to work as the derived conditions
- but those functions are only actually accessed after all this
- Therefore: putting `RegisterDerivedConditions()` above reading the savefile, is perfectly fine

...but, it's not. quite
- having done that, in both `Load()` and `New()`, I tested.
- Loading and re-saving (even by launching then re-loading from midair, triggers a file write) scrubs the new file of the `reputation: __` conditions.
- Loading from before the reputation-changing mission now works. When I land on Tarazed the government reputation is correctly set by the mission
- **But** loading from the broken save after the mission action, when `reputation with` and `condition .. reputation: _` are out of sync, does *not* set the in-game reputation from the savefile condition, it apparently just ignores the savefile condition

I don't understand how this can happen. Therefore I've missed something.
- A mission action successfully updates the in-game reputation
- Loading from the savefile successfully sees there's already derived-provider for `reputation: ` and doesn't add spurious entries
- *but* either doesn't write into the provider properly (how is this different from a mission action?!)
- or they're being re-overwritten from the savgame's `"reputation with"` block

I've tested with a PrintTrace that when loading the file, `"reputation with"` is read from the file before all the `"conditions"`

**Comment by leklachu** [comment]
Oh! Loading `"reputation with"` sets through `reputationsChanges.emplace_back(..)`, but `"conditions"` sets through `gov->SetReputation(value)`. I bet the former is only being applied after the load in the call to `ApplyChanges()`

I'll look later. But if so that explains.
- ...then I need to understand *why* ApplyChanges is being called rather than just setting the reputations directly on load, and if it needs to be that way (and what implication that has for other derived providers)

**Comment by leklachu** [comment]
I have made a draft PR that, I believe, fixes the bug, both for saves before it matters and saves that now have out-of-sync reputations.
- I haven't yet checked it to my satisfaction
- But I'm pretty confident of the important bits now
- Making this PR so you can look and see if it's a good way, because I think this bug needs fixing ASAP

https://github.com/endless-sky/endless-sky/pull/7224

---

I think I've understood enough now to write @petervdmeer's requested warning, but with this fix I think maybe it stops being useful.
- let me know, and I'll have a stab at it

**Comment by petervdmeer** [comment]
> I think I've understood enough now to write @petervdmeer's requested warning, but with this fix I think maybe it stops being useful.

Your draft PR looks good, please proceed with it.

I added another PR (https://github.com/endless-sky/endless-sky/pull/7226) which introduces the warning, because I feel the warning can also help to avoid similar bugs in the future. (So your PR fixes a bug today, my PR with the warning might help to prevent similar bugs in the future.)

**Comment by leklachu** [comment]
So, I think the main issue of this is solved with https://github.com/endless-sky/endless-sky/pull/7224, but I'd like to leave the issue open to remember to fix new games. As far as I saw, currently new games load any conditions directly into the key/value store, not through the `set` interface.

- If a new game adds conditions that should be derived, they will probably nullify the derived conditions, but should at least warn, once https://github.com/endless-sky/endless-sky/pull/7226 is merged
- Simply registering the derived conditions first, as with the loaded games, won't help: in this case the added primary condition will force itself in anyway, and not trigger any warning.

A fix is probably in two parts:

- change the `ContitionSet::Apply` to use ConditionStore's proper `Set` function
- put `RegisterDerivedConditions` above `start..Apply`

Obviously it needs to be checked carefully though, not relying on my half-understanding. I'll aim to look at this myself, but can't make any promises in the next week or so.

**Comment by petervdmeer** [comment]
> * change the `ContitionSet::Apply` to use ConditionStore's proper `Set` function

Does the override of the `operator[]` in https://github.com/endless-sky/endless-sky/blob/9ef4e92ca076d5f1ad9d5fcbc8ba1bc47cd51c01/source/ConditionsStore.h#L171 help to address this problem?

> * put `RegisterDerivedConditions` above `start..Apply`

Agreed, I added this to https://github.com/endless-sky/endless-sky/pull/7226 , because it seems to be a good idea.

**Comment by leklachu** [comment]
> Does the override of the operator[] ... help to address this problem?

Apparently, yes. Took me a while to realise what's going on. Is there a reason to have two separate ways of getting and setting a condition's value: `GetEntry()` and `Set()`, and `conditions["..."] = <value>` (with overloaded `[]` and `=`); or is that just an artefact of how it's been developed at different times?

After banging my head to logic the logic, I just tested it. Made a game start that sets reputation: currently it fails as expected. Swapped the lines in PlayerInfo.cpp (`RegisterDerivedConditions` and `start..Apply`), and it works. I'm calling it a night! (Though I'm also satisfied now that I think I understand the logic and it checks out.)

Since you've swapped the lines in #7226, that should solve it, and this issue can be closed once that is merged.

**Comment by Amazinite** [comment]
While more improvements could perhaps be done, it looks like #7224 was sufficient for fixing this issue, so we can close this now.

**Comment by leklachu** [comment]
eh, kind of. Fixed the issue that spawned this bug but not the equivalent (but less likely) one that got shown up by this, that should be fixed with #7226
